library/file.lisp

```diff
@@ -0,0 +1,724 @@
+                        path))
+
+  ;;
+  ;; Handling directory behavior that depends on emptiness
+  ;;
+  
+  (declare empty? ((Into :a Pathname) => :a -> (Result FileError Boolean)))
+  (define (empty? path)
+    "Checks whether a directory is empty."
+    (%if-directory-path (fn (p)
+                          (pure (lisp Boolean (p)
+                                  (cl:null (cl:directory (cl:merge-pathnames uiop:*wild-directory* p))))))
+                        path))
+
+  (declare remove-directory! ((Into :a Pathname) => :a -> (Result FileError :a)))
+  (define (remove-directory! path)
+    "Deletes an empty directory."
+    (let p = (the Pathname (into path)))
+    (lisp (Result FileError :a) (p)
+      (%handle-file-function (uiop:delete-empty-directory p))))
+  
+  (declare remove-directory-recursive! ((Into :a Pathname) => :a -> (Result FileError Unit)))
+  (define (remove-directory-recursive! path)
+    "Deletes a target directory recursively. Equivalent to `rm -r`. Errors if the path is not a directory."
+    (%if-directory-path (fn (p)
+                          (lisp (Result FileError Unit) (p)
+                            (%handle-file-function (uiop:delete-directory-tree p :validate cl:t))))
+                        path))
+  
+  (declare system-relative-pathname ((Into :a String) => :a -> String -> (Result FileError Pathname)))
+  (define (system-relative-pathname system-name name)
+    "Generates a system-relative-pathname for a given filename or path. This is a wrapper for `asdf:system-relative-pathname`. `Name` will likely be an empty string unless a subdirectory or filename is specified."
+    (lisp (Result FileError Pathname) (system-name name)
+      (cl:handler-case (Ok (asdf:system-relative-pathname system-name name))
+        (cl:error (c) (Err (LispError c)))))))
+
+;;;
+;;; Basic File Operations
+;;;
+
+(coalton-toplevel
+
+  (declare copy! ((Into :a Pathname) (Into :b Pathname) => :a -> :b -> (Result FileError Unit)))
+  (define (copy! input output)
+    "Copies a file to a new location."
+    (do
+     (let in = (the Pathname (into input)))
+     (let out = (the Pathname (into output)))
+      (if (not (file-pathname? in))
+          (Err (PathError "Invalid input for copying, path is not a file:" in))
+          (lisp (Result FileError :c) (in out)
+            (%handle-file-function (uiop:copy-file in out))))))
+
+  (declare delete-file! ((Into :a Pathname) => :a -> (Result FileError Unit)))
+  (define (delete-file! path)
+    "Deletes a given file if the file exists."
+    (do
+     (let p = (the Pathname (into path)))
+     (lisp (Result FileError Unit) (p)
+       (%handle-file-function (uiop:delete-file-if-exists p))))))
+
+;;;
+;;; FStreams, FileStreams, and options
+;;;
+
+(coalton-toplevel
+
+  (repr :native cl:file-stream)
+  (define-type (FileStream :a)
+    "Represents a file stream, using `cl:file-stream`.")
+
+  (repr :enum)
+  (define-type IfExists
```

Comment:
I would prefer another name personally, but I'm not going to die on the hill.

Reply:
Would `FileExists` be better? I have no real opinions about the name for this.